UniversalAlgebra/Structures/Homs/Records.lagda

```diff
@@ -9,86 +9,71 @@ author: [agda-algebras development team][]
-open import Level                  using    ( Level ;  Lift
-                                            ; lift  ;  lower    )
-                                   renaming ( zero  to ℓ₀       )
-open import Function.Base          using    ( _∘_   ;  id       )
-open import Relation.Binary        using    ( IsEquivalence     )
+open import Agda.Builtin.Equality  using ( _≡_ ; refl )
+open import Agda.Primitive         using ( _⊔_ ; lsuc ) renaming ( Set to Type )
+open import Data.Product           using ( _,_ ; Σ ; _×_ ; Σ-syntax ) renaming ( proj₁ to fst ; proj₂ to snd )
+open import Level                  using ( Level ;  Lift ; lift ; lower ) renaming ( zero  to ℓ₀ )
+open import Function.Base          using ( _∘_ ; id )
+open import Relation.Binary        using ( IsEquivalence )
 open import Relation.Binary.PropositionalEquality
-                                   using    ( cong  ; module ≡-Reasoning
-                                            ; sym   ; trans )
-
+                                   using ( cong ; module ≡-Reasoning ; sym ; trans )
 
 
-
-open import Overture.Preliminaries    using ( ℓ₁ ; ∣_∣ ; ∥_∥ ; _⁻¹ ; _∙_ ; 𝑖𝑑 ; Π ; Π-syntax)
-open import Overture.Inverses         using ( IsInjective ; IsSurjective ; Image_∋_)
-open import Relations.Discrete        using ( ker ; kerlift )
-open import Relations.Quotients       using ( Equivalence ; Quotient
-                                            ; 0[_]Equivalence ; ker-IsEquivalence
-                                            ; kerlift-IsEquivalence ; ⟪_⟫ ; ⌞_⌟
-                                            ; ⟪_∼_⟫-elim ; _/_ )
-open import Relations.Extensionality  using ( swelldef )
-open import Structures.Records        using ( signature ; structure ; Lift-struc ; compatible )
-open import Structures.Examples       using ( Sig∅ )
+open import Overture.Preliminaries   using ( ℓ₁ ; ∣_∣ ; ∥_∥ ; _⁻¹ ; _∙_ ; 𝑖𝑑 ; Π ; Π-syntax)
+open import Overture.Inverses        using ( IsInjective ; IsSurjective ; Image_∋_)
+open import Relations.Discrete       using ( ker ; kerlift )
+open import Relations.Quotients      using ( Equivalence ; Quotient ; 0[_]Equivalence
+                                           ; ker-IsEquivalence ; kerlift-IsEquivalence
+                                           ; ⟪_⟫ ; ⌞_⌟ ; ⟪_∼_⟫-elim ; _/_ )
+open import Relations.Extensionality using ( swelldef )
+open import Structures.Records       using ( signature ; structure ; Lift-Struc
+                                           ; Lift-Strucˡ ; compatible ; siglev₁)
+open import Structures.Examples      using ( Sig∅ )
 
 
 open structure
 open signature
 
-private variable 𝐹 𝑅 : signature
+private variable
+ 𝓞₀ 𝓥₀ 𝓞₁ 𝓥₁ : Level
+ 𝐹 : signature 𝓞₀ 𝓥₀
+ 𝑅 : signature 𝓞₁ 𝓥₁
+ α ρᵃ β ρᵇ γ ρᶜ : Level
+
 
-module _ {α ρᵃ : Level}
-         (𝑨 : structure 𝐹 {α}𝑅 {ρᵃ})
-         {β ρᵇ : Level}
-         (𝑩 : structure 𝐹 {β}𝑅 {ρᵇ}) where
+module _ (𝑨 : structure 𝐹 𝑅 {α}{ρᵃ})
```

Comment:
If you find that you **have** to specify implicits often, then it's usually a sign that they should be explicit.

Reply:
Got it, thanks.

Reply:
I think for a type like structure (which will often be the return type of a function), it may be reasonable to make level parameters implicit because sometimes we can omit them (e.g., when a structure is returned by a function).  However, there are also many cases where a structure will be an *input* argument to a function, and then we'll most likely have to specify the level parameters ourselves.

...so, once again, I'm not sure if we should make level parameters explicit.

Reply:
Well, I'll leave it as is for now... but we might opt to make the level parameters explicit later.

Reply:
It's tricky. When you have a `structure`, you can always infer its level. But if you're working over an arbitrary structure, then you can never do so, and must always specify the levels. So it depends on whether you mostly use them abstractly or concretely.